### PR TITLE
[1.9]Random block tick event

### PR DESCRIPTION
**Pull request by Artem226**
Hello, MinecraftForge team!

I've added a new event, that has a lot of uses - RandomTickEvent. 
This event is called when World is update blocks( World.updateBlocks()).

To prevent a huge performance fall, I've added a RandomTickEventRegistry to mark which blocks send the event. If a block isn't registred in this registry, it won't send this event.

Also event won't be sent, if block block.getTickRandomly().

I hope you merge this PR.

### CODE REVIEW

**Comment by GotoLink** [comment]
The random parameter in the event is kinda useless.
You may want to check whether the block state changed after the event is posted, or provide a setter for a possible 'new' block state.
Why is the registry using Class instead of instances ? Did you check the performance cost on that ?

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 1]
```diff
@@ -975,4 +975,31 @@ public static void onEmptyClick(EntityPlayer player, EnumHand hand)
     {
         MinecraftForge.EVENT_BUS.post(new PlayerInteractEvent.RightClickEmpty(player, hand));
     }
+
+    public static void onBlockRandomTick(World worldIn, BlockPos pos, IBlockState state, Random random)
+    {
+    	if(BlockEvent.RandomTickEvent.RandomBlockTickEventRegistry.isEventRegistredForBlock(state.getBlock()))
+    	{
+    		BlockEvent.RandomTickEvent rte = new BlockEvent.RandomTickEvent(worldIn, pos, state);
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: No tabs

**Comment by williewillus** [comment]
I don't get the point of this registry thing

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 2]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+		 * 
+		 * <strong>!!!WARNING!!! Please, don't register too many blocks, because it may reduce Minecraft's performance!</strong><br>
+		 * <strong>Before you register the event for a block, try to look for alternative ways!</strong>
+		 * <br><br>
+		 * 
+		 * <strong> !!!Fires only if the block.getTickRandomly() is true!!!
+		 * 
+		 *
+		 */
+		public static abstract class RandomBlockTickEventRegistry
```
Comment (Artem226) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: @williewillus Here. To register event use registerEventForBlock(Block block)

**Comment by Artem226** [comment]
@williewillus This registry is used to mark blocks before updating which will be called event.

**Comment by williewillus** [comment]
Yes, but why does forge need to have this? Modders should do this themselves

**Comment by williewillus** [comment]
The way I'm seeing it you're trying to "avoid" events being fired unless they're registered
But that effort is pointless - Render events and tickevents are fired many times every second already so this just seems like premature optimization

**Comment by Artem226** [comment]
@williewillus 

1. There are a rule "randomTickSpeed" in Minecraft and it is set to 3 by default, it means that block update isn't causing every tick.

2. There are block.getTickRandomly() check.
`if (block.getTickRandomly())
                                {
                                    net.minecraftforge.common.ForgeHooks.onBlockRandomTick(this, new BlockPos(k1 + j, i2 + extendedblockstorage.getYLocation(), l1 + k), iblockstate, this.rand);
                                }`

**Comment by williewillus** [comment]
none of those 2 statements explained why you need this "registry" thing

**Comment by gigaherz** [comment]
The random ticking already causes the block to get its `randomTick` method called, which by default gets forwarded to the `updateTick` method. What would the benefit be, from having an extra event?

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 3]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+     * <br><br>
+     * <strong>!!!WARNING!!! If the block isn't registred by {@link RandomTickEvent.RandomTickEventRegistry}, the event won't be sent!!!</strong>
+     */
+    @Cancelable
+    public static class RandomTickEvent extends BlockEvent
+    {
+    	private IBlockState state;
+    	
+		public RandomTickEvent(World world, BlockPos pos, IBlockState state) {
+			super(world, pos, state);
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: the indentation here is kinda wacky

File: src/test/java/net/minecraftforge/test/RandomBlockTickTest.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	public static final String MODID = "RandomBlockTickEventTest";
+    public static final String VERSION = "1.0";
+    
+    public static final boolean ENABLE = false;
+    
+    @EventHandler
+    public void preInit(FMLPreInitializationEvent event)
+    {
+    	if(!ENABLE)return;
+    	FMLLog.info("{Artem226} Preinit test mod");
```
Comment (williewillus) [inline src/test/java/net/minecraftforge/test/RandomBlockTickTest.java]: put the actual modid in there if you're going to do that

**Comment by PaleoCrafter** [comment]
I think the obvious benefit of an extra event is that mods can intercept other mods'/Vanilla's blocks ticking and do something there. One use case might be speeding up the decay of leaves (i.e. one block receives the random tick and the event handler propagates it to surrounding leaves).

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 5]
```diff
@@ -975,4 +975,31 @@ public static void onEmptyClick(EntityPlayer player, EnumHand hand)
...
+    		MinecraftForge.EVENT_BUS.post(rte);
+    		if(!rte.isCanceled())
+    		{
+    			state.getBlock().randomTick(worldIn, pos, state, random);
+    			BlockEvent.RandomTickEvent.Post post = new BlockEvent.RandomTickEvent.Post(worldIn, pos, state);
+    			MinecraftForge.EVENT_BUS.post(post);
+    		}
+    		else
+    		{
+    			if(!(rte.getState() == state))
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: why not just `if(rte.getState() != state)`

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java (hunk 3 above)
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: Not to mention the newline issue with those brackets. Some get newlines and others don't.

**Comment by LexManos** [comment]
Why do we care for this event?
Is it worth the performance hit?
If you want to stop 'random' updates then it would be better to do it at the chunk level, or something This is to deep into the tick cycle.

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 6]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+     * 
+     * If this event is canceled, the block won't do updates, but if you changed the IBlockState variable, the block in the world would be changed to the new one.
+     * <br>
+     * This event is called on the {@link MinecraftForge.#EVENT_BUS}
+     * <br>
+     * 
+     * This event is {@link Cancelable}
+     * 
+     * <br><br>
+     * <strong>!!!WARNING!!! If the block isn't registred by {@link RandomTickEvent.RandomTickEventRegistry}, the event won't be sent!!!</strong>
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: Shouldn't this be a documentation note?

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 7]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+			this.state = state;
+		}
+		
+		public IBlockState getState()
+		{
+			return state;
+		}
+		
+		/**
+		 * Does the same thing as {@link RandomTickEvent}, but it isn't cancelable and it is called just after the block updated. Sure, if the block has been registred in {@link RandomBlockTickEventRegistry}
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: Pretty sure this is considered a bit vaguely documented, could have copy and pasted docs.

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 8]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+				super(world, pos, state);
+			}
+		}
+
+		/**
+		 * A class to register {@link BlockEvent.RandomTickEvent} behavior for the block
+		 * 
+		 * <br><br>
+		 * 
+		 * <strong>!!!WARNING!!! Please, don't register too many blocks, because it may reduce Minecraft's performance!</strong><br>
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: Same as above

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 9]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+		 * <strong>Before you register the event for a block, try to look for alternative ways!</strong>
+		 * <br><br>
+		 * 
+		 * <strong> !!!Fires only if the block.getTickRandomly() is true!!!
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: Also odd

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 10]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+			{
+				return blockList.contains(blockClazz);
+			}
+			
+			public static boolean isEventRegistredForBlock(Block block)
+			{
+				return blockList.contains(block.getClass());
+			}
+			
+			public static void registerEventForBlock(Class<? extends Block> blockClazz)
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: I would think this name is backwards, as you'd be registering a block for event interaction.

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 11]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+			{
+				if(isEventRegistredForBlock(blockClazz))
+				{
+					FMLLog.warning("Random Block Tick Event has just been tried to be added to the registry, but it has been already registred!");
+					return;
+				}
+				blockList.add(blockClazz);
+			}
+			
+			public static void registerEventForBlock(Block block)
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: ditto with above

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 12]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+				return blockList;
+			}
+
+			public static void setBlockList(List<Class<? extends Block>> blockList) {
+				RandomBlockTickEventRegistry.blockList = blockList;
+			}
+
+			private static List<Class<? extends Block>> blockList = new ArrayList<Class<? extends Block>>();
+			
+			public static boolean isEventRegistredForBlock(Class<? extends Block> blockClazz)
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: isBlcokRegistered my be a better name, this seems a bit large

File: src/main/java/net/minecraftforge/event/world/BlockEvent.java [hunk 13]
```diff
@@ -233,4 +237,112 @@ public NeighborNotifyEvent(World world, BlockPos pos, IBlockState state, EnumSet
...
+		 * <strong>Before you register the event for a block, try to look for alternative ways!</strong>
+		 * <br><br>
+		 * 
+		 * <strong> !!!Fires only if the block.getTickRandomly() is true!!!
+		 * 
+		 *
+		 */
+		public static abstract class RandomBlockTickEventRegistry
+		{
+			private RandomBlockTickEventRegistry(){}
```
Comment (Spartan322) [inline src/main/java/net/minecraftforge/event/world/BlockEvent.java]: if it's abstract, is a private constructor needed?

**Comment by Spartan322** [comment]
I feel like this event was thought out enough, the standards here may have also gotten the same treatment.

**Comment by Artem226** [comment]
About performance. There are 2 screenshots below. One shows the MC info when the event and it's checking stuff is disabled, the second when enabled.

**Comment by williewillus** [comment]
How about without the registry?

**Comment by Artem226** [comment]
@williewillus Ok, I'll try.

**Comment by Spartan322** [comment]
If we are gonna use a registry, why not just make it a functor? It would cut down on performance anyway.

**Comment by diesieben07** [comment]
How do screenshots show anything about performance? Am I missing something?

**Comment by Artem226** [comment]
@diesieben07 You can see fps, memory usage.

**Comment by covers1624** [comment]
@Artem226 FPS and Mem usage mean nothing for performance in this instance. How about total time per tick? Without event, With event with registry, With event without registry.

**Comment by williewillus** [comment]
use the vanilla profiler

**Comment by Artem226** [comment]
TPS Without event:
Without event

TPS With event & registry
With event & registry

**Comment by covers1624** [comment]
Interesting, can you try this test again but in more sterile conditions (no mobs spawning / things changed from test to test except the event being on or off), it is difficult to determine if the increase in tick time is because of more entities things happening.

**Comment by williewillus** [comment]
Yes, those testing conditions are pretty poor

**Comment by gigaherz** [comment]
You could use a process like this: in a voidworld, set the randomTickSpeed to 0 (disables random ticks, default is 3), plant "large" field of crops, and move yourself to a location that can see all the crops, then set the tick rate to a high value until the crops finished growing.
You can use a profiler to measure the overall "cpu time" used by this test, run it more than once (5-10 times for each version) and average the values. 
The important thing is that you have the same starting conditions, and same settings.

**Comment by williewillus** [comment]
^ you can use /fill to plant the crops and reset the test for another trial run

**Comment by Artem226** [comment]
@gigaherz , @williewillus , Ok, I've done that you've asked.

I've created a superflat world, set difficulty to 0(peacefull) and make a field from fire blocks(because I've subscribed the event on them). 

Without event:
Without event
 
With event(some fire blocks have already turned into diamond blocks, because I've set the new blockstate in the event handler to indicate that event works):
With event

**Comment by williewillus** [comment]
Generate some text reports with /debug

**Comment by covers1624** [comment]
The testing conditions are still not good enough imo. Create a void world, and do it exactly as @gigaherz has stated and that will give us a good indication of how much a performance hit this is. Until i see some good debug stuff its a -1 from me due to the lack of evidence that the event is harmless.

**Comment by liach** [comment]
But you need to actually make a event handler that listens to this event. Preferably, this event handler should do some real things that affects performance, like running a loop. Simply posting such events are not convincible.

**Comment by blay09** [comment]
The fact that there needs to be such intense performance testing makes me really not want this event, especially since there still hasn't been any mention of an actual use case apart from Fast Leaf Decay; speaking of which, this event would fire much more rapidly in areas with many, many leaves like jungles I assume?

**Comment by gigaherz** [comment]
I have been wondering: how many blocks do actually random-tick per second on average, with default random tick speed? If it's just a few dozen, then the performance shouldn't really matter -- we have events fired more often than that with no issues. If it is in the thousands, that'd be different, but I don't think that's how random ticking works? IIRC, random ticking selects a number of blocks per tick, up to the number specified in the random tick speed gamerule? If so, then server cpu usage wouldn't me so much of a concern, IMHO.

**Comment by Artem226** [comment]
There are a function "updateBlocks" in class WorldServer. You may understand block updating logic from it.   
```java
protected void updateBlocks()
    {
        this.playerCheckLight();

        if (this.worldInfo.getTerrainType() == WorldType.DEBUG_WORLD)
        {
            Iterator<Chunk> iterator1 = this.thePlayerManager.getChunkIterator();

            while (iterator1.hasNext())
            {
                ((Chunk)iterator1.next()).onTick(false);
            }
        }
        else
        {
            int i = this.getGameRules().getInt("randomTickSpeed");
            boolean flag = this.isRaining();
            boolean flag1 = this.isThundering();
            this.theProfiler.startSection("pollingChunks");

            for (Iterator<Chunk> iterator = getPersistentChunkIterable(this.thePlayerManager.getChunkIterator()); iterator.hasNext(); this.theProfiler.endSection())
            {
                this.theProfiler.startSection("getChunk");
                Chunk chunk = (Chunk)iterator.next();
                int j = chunk.xPosition * 16;
                int k = chunk.zPosition * 16;
                this.theProfiler.endStartSection("checkNextLight");
                chunk.enqueueRelightChecks();
                this.theProfiler.endStartSection("tickChunk");
                chunk.onTick(false);
                this.theProfiler.endStartSection("thunder");

                if (this.provider.canDoLightning(chunk) && flag && flag1 && this.rand.nextInt(100000) == 0)
                {
                    this.updateLCG = this.updateLCG * 3 + 1013904223;
                    int l = this.updateLCG >> 2;
                    BlockPos blockpos = this.adjustPosToNearbyEntity(new BlockPos(j + (l & 15), 0, k + (l >> 8 & 15)));

                    if (this.isRainingAt(blockpos))
                    {
                        DifficultyInstance difficultyinstance = this.getDifficultyForLocation(blockpos);

                        if (this.rand.nextDouble() < (double)difficultyinstance.getAdditionalDifficulty() * 0.05D)
                        {
                            EntityHorse entityhorse = new EntityHorse(this);
                            entityhorse.setType(HorseArmorType.SKELETON);
                            entityhorse.setSkeletonTrap(true);
                            entityhorse.setGrowingAge(0);
                            entityhorse.setPosition((double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ());
                            this.spawnEntityInWorld(entityhorse);
                            this.addWeatherEffect(new EntityLightningBolt(this, (double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ(), true));
                        }
                        else
                        {
                            this.addWeatherEffect(new EntityLightningBolt(this, (double)blockpos.getX(), (double)blockpos.getY(), (double)blockpos.getZ(), false));
                        }
                    }
                }

                this.theProfiler.endStartSection("iceandsnow");

                if (this.provider.canDoRainSnowIce(chunk) && this.rand.nextInt(16) == 0)
                {
                    this.updateLCG = this.updateLCG * 3 + 1013904223;
                    int j2 = this.updateLCG >> 2;
                    BlockPos blockpos1 = this.getPrecipitationHeight(new BlockPos(j + (j2 & 15), 0, k + (j2 >> 8 & 15)));
                    BlockPos blockpos2 = blockpos1.down();

                    if (this.canBlockFreezeNoWater(blockpos2))
                    {
                        this.setBlockState(blockpos2, Blocks.ice.getDefaultState());
                    }

                    if (flag && this.canSnowAt(blockpos1, true))
                    {
                        this.setBlockState(blockpos1, Blocks.snow_layer.getDefaultState());
                    }

                    if (flag && this.getBiomeGenForCoords(blockpos2).canRain())
                    {
                        this.getBlockState(blockpos2).getBlock().fillWithRain(this, blockpos2);
                    }
                }

                this.theProfiler.endStartSection("tickBlocks");

                if (i > 0)
                {
                    for (ExtendedBlockStorage extendedblockstorage : chunk.getBlockStorageArray())
                    {
                        if (extendedblockstorage != Chunk.NULL_BLOCK_STORAGE && extendedblockstorage.getNeedsRandomTick())
                        {
                            for (int i1 = 0; i1 < i; ++i1)
                            {
                                this.updateLCG = this.updateLCG * 3 + 1013904223;
                                int j1 = this.updateLCG >> 2;
                                int k1 = j1 & 15;
                                int l1 = j1 >> 8 & 15;
                                int i2 = j1 >> 16 & 15;
                                IBlockState iblockstate = extendedblockstorage.get(k1, i2, l1);
                                Block block = iblockstate.getBlock();
                                this.theProfiler.startSection("randomTick");

                                if (block.getTickRandomly())
                                {
                                    net.minecraftforge.common.ForgeHooks.onBlockRandomTick(this, new BlockPos(k1 + j, i2 + extendedblockstorage.getYLocation(), l1 + k), iblockstate, this.rand);
                                }

                                this.theProfiler.endSection();
                            }
                        }
                    }
                }
            }

            this.theProfiler.endSection();
        }
    }
```


+ updateLCG declaration from World class
```java
protected int updateLCG = (new Random()).nextInt();
```

**Comment by gigaherz** [comment]
You didn't need to paste that, I'd have been able to read it as soon as I got home form work. XD

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 14]
```diff
@@ -975,4 +975,31 @@ public static void onEmptyClick(EntityPlayer player, EnumHand hand)
...
         MinecraftForge.EVENT_BUS.post(new PlayerInteractEvent.RightClickEmpty(player, hand));
     }
+
+    public static void onBlockRandomTick(World worldIn, BlockPos pos, IBlockState state, Random random)
+    {
+    	if(BlockEvent.RandomTickEvent.RandomBlockTickEventRegistry.isBlockRegistred(state.getBlock()))
+    	{
+    		BlockEvent.RandomTickEvent rte = new BlockEvent.RandomTickEvent(worldIn, pos, state);
+    		MinecraftForge.EVENT_BUS.post(rte);
+    		if(!rte.isCanceled())
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: Still tabs...

**Comment by Actuarius** [comment]
@williewillus added labels [Feature]

**Comment by Zetal911** [comment]
If performance is still an issue because of the nested for loops, you could move the event hook outside of the loops, just about the first 'ExtendedBlockStorage' loop, and pass the chunk into the event. The event could be cancelled. 

That way, people can still accomplish what they need to do by just rewriting the small segment of code that's in the for loops from within their event handler, and the people that don't need to change anything won't be affected by the performance hit because it's only firing the forge event once every chunk tick, instead of a bazillion times every chunk tick.

Unless I'm just lost and confused, which is always a possibility.

**Comment by mezz** [comment]
Based on what's been posted here, this hits performance pretty hard.
If you want to update this for a newer Forge version and try to fix performance go ahead, but this one is getting closed.

**Pull request closed without merging**